### PR TITLE
[FLINK-23715][parquet] Support for reading fields that do not exist in Parquet files

### PR DESCRIPTION
**Pull request by Tartarus0zm**
## What is the purpose of the change

Support for reading fields that do not exist in Parquet files

## Brief change log

Record non-existent columns and fill in null when reading

## Verifying this change

add test ParquetColumnarRowInputFormatTest#testProjectionReadUnknownField

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by Tartarus0zm** [comment]
@lirui-apache  please take a look, if you have time, thanks

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetColumnarRowInputFormatTest.java [hunk 1]
```diff
@@ -253,6 +253,47 @@ public void testProjection() throws IOException {
...
+                createTempParquetFile(
+                        TEMPORARY_FOLDER.newFolder(), PARQUET_SCHEMA, records, rowGroupSize);
+
+        // test reader
+        LogicalType[] fieldTypes =
+                new LogicalType[] {
+                    new DoubleType(), new TinyIntType(), new IntType(), new VarCharType()
+                };
+        ParquetColumnarRowInputFormat<FileSourceSplit> format =
+                new ParquetColumnarRowInputFormat(
```
Comment (lirui-apache) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetColumnarRowInputFormatTest.java:275]:
```suggestion
                new ParquetColumnarRowInputFormat<>(
```

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetColumnarRowInputFormatTest.java [hunk 2]
```diff
@@ -253,6 +253,47 @@ public void testProjection() throws IOException {
...
+        AtomicInteger cnt = new AtomicInteger(0);
+        forEachRemaining(
+                format.createReader(
+                        EMPTY_CONF, new FileSourceSplit("id", testPath, 0, Long.MAX_VALUE)),
+                row -> {
+                    int i = cnt.get();
+                    assertEquals(i, row.getDouble(0), 0);
+                    assertEquals((byte) i, row.getByte(1));
+                    assertEquals(i, row.getInt(2));
+                    assertEquals(true, row.isNullAt(3));
```
Comment (lirui-apache) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetColumnarRowInputFormatTest.java:292]:
```suggestion
                    assertTrue(row.isNullAt(3));
```

**Review by lirui-apache — COMMENTED** [review]
Thanks @Tartarus0zm for working on this topic. I have left some comments.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetVectorizedInputFormat.java [hunk 3]
```diff
@@ -361,8 +377,11 @@ private boolean nextBatch(ParquetReaderBatch<T> batch) throws IOException {
 
             int num = (int) Math.min(batchSize, totalCountLoadedSoFar - rowsReturned);
             for (int i = 0; i < columnReaders.length; ++i) {
-                //noinspection unchecked
```
[the remark this answers is not in the document]
Comment (Tartarus0zm) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetVectorizedInputFormat.java:364]: Deleted by mistake, added back

**Comment by Tartarus0zm** [comment]
@lirui-apache   thanks for your review,  I have solved the problem you left, please take a look again.

**Review by lirui-apache — APPROVED** [review]
+1 pending CI. Thanks for updating.

**Comment by lirui-apache** [comment]
Hey @Tartarus0zm , could you check whether the CI failure is related? If you have your own azure pipeline and it has passed, you can also post the link here.

**Comment by Tartarus0zm** [comment]
> Hey @Tartarus0zm , could you check whether the CI failure is related? If you have your own azure pipeline and it has passed, you can also post the link here.

@lirui-apache  Thank you for your reminder, I think this FAILED is not related to my code, I will trigger CI again

**Comment by lirui-apache** [comment]
@flinkbot run azure

**Comment by Tartarus0zm** [comment]
@flinkbot run azure

**Comment by Tartarus0zm** [comment]
@flinkbot run azure

**Comment by lirui-apache** [comment]
Thanks @Tartarus0zm for updating. I'll merge this once the 1.14 release branch is cut.

**Pull request closed without merging**